internal/loader/configuration_client_manager.go

```
@@ -0,0 +1,433 @@
+			return nil, err
+		}
+		if staticClient, err = azappconfig.NewClient(*provider.Spec.Endpoint, manager.credential, clientOptionWithModuleInfo); err != nil {
+			return nil, err
+		}
+		manager.endpoint = *provider.Spec.Endpoint
+	}
+
+	manager.validDomain = getValidDomain(manager.endpoint)
+	manager.StaticClientWrappers = []*ConfigurationClientWrapper{{
+		Endpoint:       manager.endpoint,
+		Client:         staticClient,
+		BackOffEndTime: metav1.Time{},
+		FailedAttempts: 0,
+	}}
+
+	return manager, nil
+}
+
+func (manager *ConfigurationClientManager) GetClients() []*ConfigurationClientWrapper {
+	currentTime := metav1.Now()
+	clients := make([]*ConfigurationClientWrapper, 0)
+	for _, clientWrapper := range manager.StaticClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	if !manager.ReplicaDiscoveryEnabled {
+		return clients
+	}
+
+	if currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) &&
+		(manager.DynamicClientWrappers == nil ||
+			currentTime.After(manager.lastFallbackClientRefresh.Time.Add(FallbackClientRefreshExpireInterval))) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.endpoint)
+		_ = manager.DiscoverFallbackClients(url.Host)
+	}
+
+	for _, clientWrapper := range manager.DynamicClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	return clients
+}
+
+func (manager *ConfigurationClientManager) RefreshClients() {
+	currentTime := metav1.Now()
+	if manager.ReplicaDiscoveryEnabled &&
+		currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.endpoint)
+		_ = manager.DiscoverFallbackClients(url.Host)
```

Should not ignore the error here.

---

Do we need to log the error?

---

DiscoverFallbackClients may return error in line 188, we should not ignore this error

---

Logging warning is fine in line 174